### PR TITLE
Fully deprecate StopCh

### PR DESCRIPTION
**Pull request by mjudeikis**
#### What type of PR is this?
/kind cleanup

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Fully remove PostStartHookContext.StopCh
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by mjudeikis** [comment]
Should be last of StopCh...

File: staging/src/k8s.io/apiserver/pkg/server/hooks.go [hunk 1]
```diff
@@ -49,11 +49,6 @@ type PreShutdownHookFunc func() error
 type PostStartHookContext struct {
 	// LoopbackClientConfig is a config for a privileged loopback connection to the API server
 	LoopbackClientConfig *restclient.Config
-	// StopCh is the channel that will be closed when the server stops.
-	//
-	// Deprecated: use the PostStartHookContext itself instead, it contains a context that
-	// gets cancelled when the server stops. StopCh keeps getting provided for existing code.
-	StopCh <-chan struct{}
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/hooks.go:56]: when did we deprecate it? Longer ago?

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/hooks.go:56]: > Fully deprecate PostStartHookContext.StopCh

Fully remove it, no?

Comment (mjudeikis) [reply staging/src/k8s.io/apiserver/pkg/server/hooks.go:56]: Its deprecated ages go. I found this message and just took on myself to do the cleaning :)

Comment (mjudeikis) [reply staging/src/k8s.io/apiserver/pkg/server/hooks.go:56]: updated release notes

Comment (mjudeikis) [reply staging/src/k8s.io/apiserver/pkg/server/hooks.go:56]: updated git commit message too

**Comment by aojea** [comment]
linter complains in
```
ERROR: staging/src/k8s.io/apiserver/pkg/server/config.go:908:24: Error return value of `c.FlowControl.Run` is not checked (errcheck)
ERROR: 			go c.FlowControl.Run(hookContext.Done())
ERROR: 			                    ^
```

**Comment by mjudeikis** [comment]
> go c.FlowControl.Run(hookContext.Done())

I was bit cautions here as we do this all around the place and hints shows this, but does not block.
I changed it now, but this changes a behaviour a bit, where, if previously it fail silently and moves on. If we check and return - it will fail server startup? 

And I was not sure if author intended this to be like this or now. It looked strange why inside `AddPostStartHook` (which is go routine on its own) we kick-off new go routine :shrug: 

Either way, changed now. Lets see what tests shows.

**Comment by mjudeikis** [comment]
wow, this just exploded in CI. Lets ee if this real or just sunday rush hour.
/retest

**Comment by aojea** [comment]
> wow, this just exploded in CI. Lets ee if this real or just sunday rush hour.
> /retest

please click on the jobs before retesting , all jobs testing failing means something is wrong.

It seems this PR is breaking the apiserver start, apiserver never gets pass the bootsrap phase https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/127341/pull-kubernetes-e2e-kind/1837906716155449344/artifacts/kind-control-plane/containers/kube-apiserver-kind-control-plane_kube-system_kube-apiserver-2179625b5987564cc3d2d04bb84be2ce671238b70a28bc5fb1ecbb7c03cd167c.log

> 2024-09-22T17:43:40.207515019Z stderr F I0922 17:43:40.207443       1 httplog.go:134] "HTTP" verb="GET" URI="/readyz" latency="2.360446ms" userAgent="kube-apiserver/v1.32.0 (linux/amd64) kubernetes/747066c" audit-ID="1fa536a2-42e2-4e9c-b34f-95af099752d0" srcIP="[::1]:59452" apf_pl="exempt" apf_fs="exempt" apf_iseats=1 apf_fseats=0 apf_additionalLatency="0s" apf_execution_time="2.039863ms" resp=500
2024-09-22T17:43:40.305483732Z stderr F I0922 17:43:40.305317       1 healthz.go:278] poststarthook/priority-and-fairness-config-consumer check failed: readyz
2024-09-22T17:43:40.30553272Z stderr F [-]poststarthook/priority-and-fairness-config-consumer failed: not finished

File: staging/src/k8s.io/apiserver/pkg/server/config.go [hunk 2]
```diff
@@ -904,9 +904,8 @@ func (c completedConfig) New(name string, delegationTarget DelegationTarget) (*G
 	const priorityAndFairnessConfigConsumerHookName = "priority-and-fairness-config-consumer"
 	if s.isPostStartHookRegistered(priorityAndFairnessConfigConsumerHookName) {
 	} else if c.FlowControl != nil {
-		err := s.AddPostStartHook(priorityAndFairnessConfigConsumerHookName, func(context PostStartHookContext) error {
-			go c.FlowControl.Run(context.StopCh)
-			return nil
+		err := s.AddPostStartHook(priorityAndFairnessConfigConsumerHookName, func(hookContext PostStartHookContext) error {
+			return c.FlowControl.Run(hookContext.Done())
```
Comment (Jefftree) [inline staging/src/k8s.io/apiserver/pkg/server/config.go:908]: this is a goroutine. We definitely should not be returning here

**Comment by Jefftree** [comment]
> linter complains in
> 
> ```
> ERROR: staging/src/k8s.io/apiserver/pkg/server/config.go:908:24: Error return value of `c.FlowControl.Run` is not checked (errcheck)
> ERROR: 			go c.FlowControl.Run(hookContext.Done())
> ERROR: 			                    ^
> ```

It's an optional linter and imo we're not making the code base worse by deprecating stopch? I think properly fixing the lint will involve changing the `FlowControl.Run()` function and interface to not return an error.

ref: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go#L371-L391

File: staging/src/k8s.io/apiserver/pkg/server/config.go (hunk 2 above)
Comment (mjudeikis) [reply staging/src/k8s.io/apiserver/pkg/server/config.go]: reverted back to original change

**Comment by mjudeikis** [comment]
> It's an optional linter and imo we're not making the code base worse by deprecating stopch? I think properly fixing the lint will involve changing the `FlowControl.Run()` function and interface to not return an error.

I was thinking same just got persuaded :D reverted now to the original change. FlowControl indeed need fixing and might need to be a separate PR.

**Comment by mjudeikis** [comment]
> please click on the jobs before retesting , all jobs testing failing means something is wrong.

As I suspected returning there will explode things. As this is a hint, once merged will create an issue to fix FlowControl as suggested by @Jefftree . One of "good first issue" ones.

**Pull request merged**